datafusion/sql/src/statement.rs

```diff
@@ -127,42 +129,66 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
+
+                        let plan = if !columns.is_empty() {
+                            let schema = self.build_schema(columns)?.to_dfschema_ref()?;
+                            if schema.fields().len() != input_schema.fields().len() {
+                                return Err(DataFusionError::Plan(format!(
                             "Mismatch: {} columns specified, but result has {} columns",
                             schema.fields().len(),
                             input_schema.fields().len()
                         )));
+                            }
+                            let input_fields = input_schema.fields();
+                            let project_exprs = schema
+                                .fields()
+                                .iter()
+                                .zip(input_fields)
+                                .map(|(field, input_field)| {
+                                    cast(
+                                        col(input_field.name()),
+                                        field.data_type().clone(),
+                                    )
+                                    .alias(field.name())
+                                })
+                                .collect::<Vec<_>>();
+                            LogicalPlanBuilder::from(plan.clone())
+                                .project(project_exprs)?
+                                .build()?
+                        } else {
+                            plan
+                        };
+
+                        Ok(LogicalPlan::CreateMemoryTable(CreateMemoryTable {
+                            name: self.object_name_to_table_reference(name)?,
+                            input: Arc::new(plan),
+                            if_not_exists,
+                            or_replace,
+                        }))
                     }
-                    let input_fields = input_schema.fields();
-                    let project_exprs = schema
-                        .fields()
-                        .iter()
-                        .zip(input_fields)
-                        .map(|(field, input_field)| {
-                            cast(col(input_field.name()), field.data_type().clone())
-                                .alias(field.name())
-                        })
-                        .collect::<Vec<_>>();
-                    LogicalPlanBuilder::from(plan.clone())
-                        .project(project_exprs)?
-                        .build()?
-                } else {
-                    plan
-                };
-
-                Ok(LogicalPlan::CreateMemoryTable(CreateMemoryTable {
-                    name: self.object_name_to_table_reference(name)?,
-                    input: Arc::new(plan),
-                    if_not_exists,
-                    or_replace,
-                }))
+
+                    None => {
+                        let schema = self.build_schema(columns)?.to_dfschema_ref()?;
```

Comment:
The problem of the `NOT NULL` default is still here since you did not change the `self.build_schema(columns)`.

This should be the default behaviour:
```rust
pub fn build_schema(&self, columns: Vec<SQLColumnDef>) -> Result<Schema> {
        let mut fields = Vec::with_capacity(columns.len());

        for column in columns {
            let data_type = self.convert_simple_data_type(&column.data_type)?;
            let prevent_null = column
                .options
                .iter()
                .any(|x| x.option == ColumnOption::NotNull);
            fields.push(Field::new(
                normalize_ident(column.name, self.options.enable_ident_normalization),
                data_type,
                !prevent_null,
            ));
        }

        Ok(Schema::new(fields))
    }
```
then, it would help if you changed the `sqllogictests`, for example, from
```
describe table_with_normalization
----
field1 Int64 NO
field2 Int64 NO
```
to
```
describe table_with_normalization
----
field1 Int64 YES
field2 Int64 YES
```

In this way, you can give the same defaults with PostgreSQL.